thrust/thrust/system/cuda/detail/partition.h

```diff
@@ -38,1041 +38,349 @@
-
-    cudaError_t status;
-    status = doit_step(NULL,
-                       temp_storage_bytes,
-                       first,
-                       stencil,
-                       selected_result,
-                       rejected_result,
-                       predicate,
-                       reinterpret_cast<size_type*>(NULL),
-                       num_items,
-                       stream);
-    cuda_cub::throw_on_error(status, "partition failed on 1st step");
-
-    size_t allocation_sizes[2] = {sizeof(size_type), temp_storage_bytes};
-    void * allocations[2]      = {NULL, NULL};
-
-    size_t storage_size = 0;
-
-    status = core::alias_storage(NULL,
-                                 storage_size,
-                                 allocations,
-                                 allocation_sizes);
-    cuda_cub::throw_on_error(status, "partition failed on 1st alias_storage");
-
-    // Allocate temporary storage.
-    thrust::detail::temporary_array<thrust::detail::uint8_t, Derived>
-      tmp(policy, storage_size);
-    void *ptr = static_cast<void*>(tmp.data().get());
-
-    status = core::alias_storage(ptr,
-                                 storage_size,
-                                 allocations,
-                                 allocation_sizes);
-    cuda_cub::throw_on_error(status, "partition failed on 2nd alias_storage");
-
-    size_type* d_num_selected_out
-      = thrust::detail::aligned_reinterpret_cast<size_type*>(allocations[0]);
-
-    status = doit_step(allocations[1],
-                       temp_storage_bytes,
-                       first,
-                       stencil,
-                       selected_result,
-                       rejected_result,
-                       predicate,
-                       d_num_selected_out,
-                       num_items,
-                       stream);
-    cuda_cub::throw_on_error(status, "partition failed on 2nd step");
-
+    // Get number of selected items
     status = cuda_cub::synchronize(policy);
-    cuda_cub::throw_on_error(status, "partition failed to synchronize");
-
-    size_type num_selected = 0;
-    if (num_items > 0)
-    {
-      num_selected = get_value(policy, d_num_selected_out);
-    }
+    CUDA_CUB_RET_IF_FAIL(status);
+    num_selected = static_cast<std::size_t>(get_value(policy, d_num_selected_out));
 
-    return thrust::make_pair(selected_result + num_selected,
-                             rejected_result + num_items - num_selected);
+    return status;
   }
+};
+
+template <typename Derived, typename InputIt, typename StencilIt, typename OutputIt, typename Predicate>
+THRUST_RUNTIME_FUNCTION std::size_t partition(
+  execution_policy<Derived>& policy,
+  InputIt first,
+  InputIt last,
+  StencilIt stencil,
+  OutputIt output,
+  Predicate predicate)
+{
+  using size_type = typename iterator_traits<InputIt>::difference_type;
+
+  size_type num_items = static_cast<size_type>(thrust::distance(first, last));
```

Comment:
question: I think distance is guaranteed to return difference type. Do we need this cast?